### PR TITLE
lxd/network/network: tell systemd-resolved we can resolve .lxd

### PR DESCRIPTION
**Pull request by ddstreet**
This allows the local system to resolve all running
lxd containers, using the (by default) .lxd domain.  This will
work only if the local system is using systemd-resolved for DNS.

Signed-off-by: Dan Streetman <ddstreet@canonical.com>

### CODE REVIEW

**Comment by stgraber** [comment]
How did you test this?

I'm not really seeing how this would work for the vast majority of our users as most are on non-systemd systems and those that are on systemd systems are likely using the snap package which doesn't let you modify the host system.

Additionally, how does this handle the case where you have multiple bridges with the same domain (`lxd`) and containers of the same name on those?

**Comment by ddstreet** [comment]
> How did you test this?

I built using the method described in doc/index.md, and ran the built lxd directly (as root).

> 
> I'm not really seeing how this would work for the vast majority of our users as most are on non-systemd systems and those that are on systemd systems are likely using the snap package which doesn't let you modify the host system.

I don't know why most users wouldn't be using systemd-resolved, as that certainly seems to be the future direction for most distros.

As far as snap isolation, lxd is able to create and configure interfaces in the host system, so it's kinda strange it wouldn't be able to also call systemd-resolve to update the DNS config.  I don't know what snap isolation lxd has though, so I will take your word for it.  That's unfortunate though as I think it would be extremely useful.

> 
> Additionally, how does this handle the case where you have multiple bridges with the same domain (`lxd`) and containers of the same name on those?

hmm, is that possible?  I just tried creating two same-named containers on a system and it failed.  If setting up multiple same-named containers is possible, what command should I use to do that?

Assuming lxd won't create multiple same-named containers on a system, then lxd will tell resolved about both (or all) nameservers, and resolved will ask both/all network's nameservers to resolve any *.lxd hostname.  The dnsmasq servers that don't know about the container will respond with NXDOMAIN and the dnsmasq server that does will respond with its address(es), meaning resolved will correctly resolve the container.

**Comment by stgraber** [comment]
You can get same named containers in different projects on the same network.
The majority of LXD users are on ChromeOS which doesn't use systemd.

The LXD snap can access the kernel which is how it manages devices, it doesn't interface with networkd/resolved or any network configuration tooling outside of the snap and would therefore be unable to configure resolved.

**Comment by stgraber** [comment]
Just thought of another problem, nested containers will also get a domain conflict as both their own network interface and the bridge they run will be on the .lxd domain.

**Comment by stgraber** [comment]
Could you convert this branch into a doc addition to `networks.md`, adding a paragraph on configuring the host system to resolve instance names with the suitable systemd-resolve/resolvectl commands?

That way those that do want this kind of integration can just configure their resolvers accordingly.

**Comment by ddstreet** [comment]
> Just thought of another problem, nested containers will also get a domain conflict as both their own network interface and the bridge they run will be on the .lxd domain.

ah, that's certainly true.

> Could you convert this branch into a doc addition to `networks.md`, adding a paragraph on configuring the host system to resolve instance names with the suitable systemd-resolve/resolvectl commands?
> 
> That way those that do want this kind of integration can just configure their resolvers accordingly.

will do, thanks.

**Comment by stgraber** [comment]
@ddstreet we're planning on releasing the next LXD on Wednesday or Thursday, would be nice if this could get updated and merged by then.

**Comment by ddstreet** [comment]
> @ddstreet we're planning on releasing the next LXD on Wednesday or Thursday, would be nice if this could get updated and merged by then.

sorry for the delay - updated to add a section to the network.md doc.  Please let me know if you would like anything changed.

Also just fyi, I threw together a small script to automate the resolved notification:
https://git.launchpad.net/~ubuntu-support-team/+git/lxdhelpers/tree/sbin/lxd-notify-resolved

and a small systemd unit file to run that script whenever the lxd service is started:
https://git.launchpad.net/~ubuntu-support-team/+git/lxdhelpers/tree/debian/lxdhelpers.lxd-notify-resolved.service

Since it will have to run outside of snapd confinement, it probably isn't useful to directly include in lxd, but thought it may be useful to mention.

File: doc/networks.md [hunk 1]
```diff
@@ -101,3 +101,32 @@ Those keys can be set using the lxc tool with:
 ```bash
 lxc network set <network> <key> <value>
 ```
+
+## Integration with systemd-resolved
+
+If the system running lxd uses systemd-resolved to perform DNS
```
Comment (stgraber) [inline doc/networks.md]: s/lxd/LXD/

File: doc/networks.md [hunk 2]
```diff
@@ -101,3 +101,32 @@ Those keys can be set using the lxc tool with:
...
+lookups, it's possible to notify resolved of the domain(s) that
+lxd is able to resolve.  This requires telling resolved the
+specific bridge(s), nameserver address(es), and dns domain(s).
+
+For example, if lxd is using the `lxdbr0` interface, get the
```
Comment (stgraber) [inline doc/networks.md]: s/lxd/LXD/

File: doc/networks.md [hunk 3]
```diff
@@ -101,3 +101,32 @@ Those keys can be set using the lxc tool with:
...
+ipv4 address with `lxd network get lxdbr0 ipv4.address` command
```
Comment (stgraber) [inline doc/networks.md]: s/lxd/lxc/

File: doc/networks.md [hunk 4]
```diff
@@ -101,3 +101,32 @@ Those keys can be set using the lxc tool with:
...
+(the ipv6 can be used instead or in addition), and the domain
+with `lxd network get lxdbr0 dns.domain` (if unset, the domain
```
Comment (stgraber) [inline doc/networks.md]: s/lxd/lxc/

File: doc/networks.md [hunk 5]
```diff
@@ -101,3 +101,32 @@ Those keys can be set using the lxc tool with:
...
+Replace `lxdbr0` with the actual bridge name, and `lxd` with the
+domain name.  Note the `~` before the domain name is important:
+it tells resolved to use this nameserver only to look up this
+domain; no matter what your actual domain name is, you should
+prefix it with `~`.  Replace `1.2.3.4` with the actual address
+of the nameserver (without the subnet netmask).
+
+This resolved configuration will persist as long as the bridge
+exists, so you must repeat this command each reboot and after
+lxd is restarted.  Also note this only works if the bridge
```
Comment (stgraber) [inline doc/networks.md]: s/lxd/LXD/

**Comment by stgraber** [comment]
@ddstreet just a few minor edits needed then we can merge this.

Thanks!

**Comment by ddstreet** [comment]
> @ddstreet just a few minor edits needed then we can merge this.
> 
> Thanks!

Fixed those lxd/LXD and lxd/lxc - also, sorry for adding a bit more, but I forgot I should also mention the 'new' resolvectl command, even if the older systemd-resolve still works it's probably good to mention what the new command syntax is.  Let me know if I should adjust anything else.

Thanks!

**Comment by stgraber** [comment]
Looks good, thanks!

**Pull request merged**